packages/react-components/react-provider/src/components/FluentProvider/useFluentProviderThemeStyleTag.ts

```diff
@@ -76,5 +86,22 @@ export const useFluentProviderThemeStyleTag = (options: Pick<FluentProviderState
     }
   }, [styleTagId, targetDocument, rule, styleElementAttributes]);
 
-  return styleTagId;
+  return { styleTagId, rule };
 };
+
+function useHandleSSRStyleElements(targetDocument: Document | undefined | null) {
+  // Using a state factory so that this logic only runs once per render
+  // Each FluentProvider can create its own style element during SSR as a slot
+  // Moves all theme style elements to document head during render to avoid hydration errors.
+  // Should be strict mode safe since the logic is idempotent.
+  React.useState(() => {
+    if (!targetDocument) {
+      return;
+    }
+
+    const themeStyleElements = targetDocument.body.querySelectorAll(`[${FUI_THEME_STYLE_ATTR}]`);
+    themeStyleElements.forEach(styleElement => {
+      targetDocument.head.append(styleElement);
+    });
+  });
+}
```

Comment:
As we discussed in offline, we don't need to query all elements, we need a specific element:

```js
function useHandleSSRStyleElements(targetDocument: Document | undefined | null, styleTagId: string) {
  // Using a state factory so that this logic only runs once per render
  // Each FluentProvider can create its own style element during SSR as a slot
  // Moves all theme style elements to document head during render to avoid hydration errors.
  // Should be strict mode safe since the logic is idempotent.
  React.useState(() => {
    if (!targetDocument) {
      return;
    }

    const themeStyleElement = targetDocument.getElementById(styleTagId);

    if (themeStyleElement) {
      targetDocument.head.append(themeStyleElement);
    }
  });
}
```